llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -3139,136 +3151,269 @@ Register SPIRVInstructionSelector::buildPointerToResource(
 }
 
-bool SPIRVInstructionSelector::selectFirstBitHigh64(Register ResVReg,
-                                                    const SPIRVType *ResType,
-                                                    MachineInstr &I,
-                                                    bool IsSigned) const {
-  Register OpReg = I.getOperand(2).getReg();
-  // 1. split our int64 into 2 pieces using a bitcast
-  unsigned count = GR.getScalarOrVectorComponentCount(ResType);
-  SPIRVType *baseType = GR.retrieveScalarOrVectorIntType(ResType);
+bool SPIRVInstructionSelector::selectFirstBitSet64Overflow(
+    Register ResVReg, const SPIRVType *ResType, MachineInstr &I,
+    Register SrcReg, unsigned BitSetOpcode, bool SwapPrimarySide) const {
+
+  unsigned ComponentCount = GR.getScalarOrVectorComponentCount(ResType);
+  SPIRVType *BaseType = GR.retrieveScalarOrVectorIntType(ResType);
+  bool ZeroAsNull = STI.isOpenCLEnv();
+  Register ConstIntZero =
+      GR.getOrCreateConstInt(0, I, BaseType, TII, ZeroAsNull);
+  unsigned LeftComponentCount = ComponentCount / 2;
+  unsigned RightComponentCount = ComponentCount - LeftComponentCount;
+  bool LeftIsVector = LeftComponentCount > 1;
+
+  // Split the SrcReg in half into 2 smaller vec registers
+  // (ie i64x4 -> i64x2, i64x2)
   MachineIRBuilder MIRBuilder(I);
-  SPIRVType *postCastT =
-      GR.getOrCreateSPIRVVectorType(baseType, 2 * count, MIRBuilder);
-  Register bitcastReg = MRI->createVirtualRegister(GR.getRegClass(postCastT));
+  SPIRVType *OpType = GR.getOrCreateSPIRVIntegerType(64, MIRBuilder);
+  SPIRVType *LeftOpType;
+  SPIRVType *LeftResType;
+  if (LeftIsVector) {
+    LeftOpType =
+        GR.getOrCreateSPIRVVectorType(OpType, LeftComponentCount, MIRBuilder);
+    LeftResType =
+        GR.getOrCreateSPIRVVectorType(BaseType, LeftComponentCount, MIRBuilder);
+  } else {
+    LeftOpType = OpType;
+    LeftResType = BaseType;
+  }
+
+  SPIRVType *RightOpType =
+      GR.getOrCreateSPIRVVectorType(OpType, RightComponentCount, MIRBuilder);
+  SPIRVType *RightResType =
+      GR.getOrCreateSPIRVVectorType(BaseType, RightComponentCount, MIRBuilder);
+
+  Register LeftSideIn = MRI->createVirtualRegister(GR.getRegClass(LeftOpType));
+  Register RightSideIn =
+      MRI->createVirtualRegister(GR.getRegClass(RightOpType));
+
+  bool Result;
+
+  // Extract the left half from the SrcReg into LeftSideIn
+  // accounting for the special case when it only has one element
+  if (LeftIsVector) {
+    auto MIB = BuildMI(*I.getParent(), I, I.getDebugLoc(),
+                       TII.get(SPIRV::OpVectorShuffle))
+                   .addDef(LeftSideIn)
+                   .addUse(GR.getSPIRVTypeID(LeftOpType))
+                   .addUse(SrcReg)
+                   // Per the spec, repeat the vector if only one vec is needed
+                   .addUse(SrcReg);
+
+    for (unsigned J = 0; J < LeftComponentCount; J++) {
+      MIB.addImm(J);
+    }
+
+    Result = MIB.constrainAllUses(TII, TRI, RBI);
```

Comment:
Can this fail, and if it does, what happens?

It looks like you're using a pattern `Result && ...` which isn't really a common LLVM pattern. LLVM prefers early exits to chaining checks (see: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code).

Maybe these should instead be rewritten as something like
```
if(!MIB.constrainAllUses(TII, TRI, RBI))
  return false;
```

Reply:
ack, yeah I was just following what the previous code does. I'm very strong proponent of early exits

(I hold the somewhat controversial opinion that you should pretty much never use `else`  even though I used it myself in this PR 🙃)